data/src/new_etl/data_utils/ndvi.py

```diff
@@ -0,0 +1,355 @@
+
+    # Define columns to include in extraction
+    include_cols = list(gdf_clean.columns)
+
+    print(f"Extracting NDVI data to {len(gdf_clean)} parcels...")
+    results = exactextract.exact_extract(
+        raster_path,
+        gdf_clean,
+        ["mean"],
+        output="pandas",
+        include_cols=include_cols,
+        strategy="raster-sequential",
+        progress=True,
+    )
+
+    return results
+
+
+@provide_metadata()
+def ndvi(primary_featurelayer: FeatureLayer) -> FeatureLayer:
+    """
+    Calculate and add NDVI features to the primary feature layer.
+
+    This function retrieves Sentinel-2 satellite imagery for the most recent summer
+    period (June-August) and the previous summer, calculates the median Normalized
+    Difference Vegetation Index (NDVI) for each period, and then adds the NDVI values
+    and year-over-year change to each parcel.
+
+    NDVI is a measure of vegetation greenness, with values ranging from -1 to 1.
+    Higher values indicate more dense, healthy vegetation.
+
+    We use the median peak NDVI over the summer months to capture when land is at
+    its greenest. Lower resolution Sentinel-2 data (10m) provides sufficient accuracy
+    for the parcel-level analysis while requiring less storage and processing time
+    than higher resolution alternatives.
+
+    Args:
+        primary_featurelayer (FeatureLayer): Input feature layer containing parcels
+
+    Returns:
+        FeatureLayer: Feature layer with added NDVI features
+
+    Note:
+        - Full processing (including data retrieval) takes approximately 10-15 minutes
+        - Extraction alone takes about 3 minutes
+        - The intermediate raster file is approximately 65MB
+        - This process typically only needs to run once per year, after summer ends
+
+    Added columns:
+        - ndvi_mean: Mean NDVI value for the parcel (current summer)
+        - ndvi_one_year_change_mean: Mean year-over-year NDVI change
+    """
+    # dir for cache files
+    temp_dir = os.path.join(os.getcwd(), "tmp")
+    os.makedirs(temp_dir, exist_ok=True)
+
+    current_summer_year = get_current_summer_year()
+    ndvi_cache_path = os.path.join(
+        temp_dir, f"ndvi_data_summer_{current_summer_year}.tif"
+    )
+
+    # Get boundary
+    boundary, bbox_coords = get_bbox_from_census_data()
+
+    # Check if we need to regenerate the NDVI data
+    if not is_cache_valid(ndvi_cache_path, current_summer_year):
+        print("Generating new NDVI data...")
+        generate_ndvi_data(ndvi_cache_path, boundary, bbox_coords)
+    else:
+        print("Using cached NDVI data")
+
+    # Get the GeoDataFrame from the feature layer
+    parcel_gdf = primary_featurelayer.gdf
+
+    # Extract NDVI values to parcels
+    results = extract_ndvi_to_parcels(ndvi_cache_path, parcel_gdf)
+
+    # Update the primary feature layer with the new columns from results
+    # (instead of creating a new one)
+    for column in results.columns:
```

Comment:
Given that results is on a cleaned version of the gdf and we're adding columns into a version that hasn't been cleaned, is there the possibility for misalignment or different row lengths here when merging in the columns?

It is probably worth making this logic a separate function and trying out some filtering on a df and making sure things merge okay.